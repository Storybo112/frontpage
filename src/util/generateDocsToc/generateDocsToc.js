const fs = require('fs');
const path = require('path');
const matter = require('gray-matter');

const NON_CONTENT_DIRECTORIES = ['assets', 'snippets', 'versions'];

function getMetadata(filePath) {
  const fileContents = fs.readFileSync(filePath, 'utf8');
  const {
    data: { navTitle, title, ...data },
  } = matter(fileContents);
  return {
    title: navTitle || title,
    ...data,
  };
}

/**
 * @param pathToDocs string - Path to docs directory, relative to project root
 *
 * @param docsRoot string - Private; Path to docs root, relative to project root (used for recursion)
 *
 * @returns TOC array
 */
module.exports = function generateDocsToc(
  pathToFiles = 'src/content/docs',
  docsRoot = pathToFiles
) {
  const files = fs.readdirSync(pathToFiles);

  if (files.includes('toc.js')) {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    const { toc } = require(path.join('../../../', pathToFiles, 'toc.js'));
    return toc;
  }

  const toc = [];
  files
    .filter((file) => !NON_CONTENT_DIRECTORIES.includes(file))
    .forEach((file) => {
      const filePath = path.join(pathToFiles, file);
      const isDirectory = fs.lstatSync(filePath).isDirectory();

      if (isDirectory) {
        const childItems = generateDocsToc(filePath, docsRoot);

        if (childItems) {
          let title;
          let redirectPath;

          const indexItemIndex = childItems.findIndex((child) => child.pathSegment === 'index');
          if (indexItemIndex !== -1) {
            const metaData = getMetadata(
              path.join(pathToFiles, file, childItems[indexItemIndex].file)
            );
            title = metaData.title;

            if (metaData.redirectTo) {
              redirectPath = path.join(file.replace(/^\d+-/, ''), metaData.redirectTo);
            }

            childItems.splice(indexItemIndex, 1);
          }

          const children = childItems
            /**
             * Items always have a `navOrder` and, when in a group, a `navGroup.order`.
             * This sorts by `navGroup.order` (falling back to `navOrder`) first, then `navOrder`.
             */
            .sort(
              (a, b) =>
                (a.navGroup?.order || a.navOrder) - (b.navGroup?.order || b.navOrder) ||
                a.navOrder - b.navOrder
            )
            .reduce(
              (
                acc,
                {
                  navGroup,
                  // Remove these from the child items, as they're not needed beyond this point
                  file: ignoreFile,
                  navOrder: ignoreNavOrder,
                  ...child
                }
              ) => {
                if (navGroup) {
                  const group = acc.find((c) => c.title === navGroup.title);
                  if (group) {
                    group.children.push(child);
                  } else {
                    acc.push({
                      pathSegment: '',
                      title: navGroup.title,
                      type: 'menu',
                      children: [child],
                    });
                  }
                } else {
                  acc.push(child);
                }
                return acc;
              },
              []
            );

          toc.push({
            pathSegment: file.replace(/^\d+-/, ''),
            ...(redirectPath ? { redirectPath } : {}),
            ...(title ? { title } : {}),
            type: 'heading',
            children,
          });
        }
      } else if (file.endsWith('.md')) {
        const metaData = getMetadata(filePath);
        toc.push({
          file,
          pathSegment: file.replace(/^(?:\d+-)?(.*)\.md$/, '$1'),
          type: 'link',
          ...metaData,
        });
      }
    });

  return toc;
};
