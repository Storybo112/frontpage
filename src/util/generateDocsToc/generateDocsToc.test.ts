import path from 'path';

import generateDocsToc from './generateDocsToc';

const prefixPath = (mockPath) => path.join('src/util/generateDocsToc/__mocks__', mockPath);

describe('generateDocsToc', () => {
  it('returns existing TOC', async () => {
    const toc = generateDocsToc(prefixPath('existing-toc'));
    const { toc: existingToc } = await import(
      path.join('../../../', prefixPath('existing-toc'), 'toc.js')
    );
    expect(toc).toEqual(existingToc);
  });

  it('sorts top-level and children items', () => {
    const toc = generateDocsToc(prefixPath('basic'));
    expect(toc).toMatchSnapshot();
  });

  it('uses `navTitle` if provided', () => {
    const toc = generateDocsToc(prefixPath('with-nav-title'));
    const topLevelItem = toc[0];
    expect(topLevelItem).toHaveProperty('title', 'Nav title');
  });

  it('generates a non-routed group, with sorted children items', () => {
    const toc = generateDocsToc(prefixPath('with-non-route-group'));
    expect(toc).toMatchSnapshot();
  });

  it('handles redirects', () => {
    const toc = generateDocsToc(prefixPath('with-redirect'));
    const topLevelItem = toc[0];
    expect(topLevelItem).toHaveProperty('redirectPath', 'parent/target');
  });

  it('parses descriptions', () => {
    const toc = generateDocsToc(prefixPath('with-description'));
    const topLevelItem = toc[0];
    expect(topLevelItem).toHaveProperty('description', 'Description');
  });
});
