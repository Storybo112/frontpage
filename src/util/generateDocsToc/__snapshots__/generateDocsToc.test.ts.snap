// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDocsToc generates a non-routed group, with sorted children items 1`] = `
[
  {
    "children": [
      {
        "pathSegment": "before",
        "title": "Before (should be first)",
        "type": "link",
      },
      {
        "children": [
          {
            "pathSegment": "parameters",
            "title": "Parameters (should be first in group)",
            "type": "link",
          },
          {
            "pathSegment": "decorators",
            "title": "Decorators (should be second in group)",
            "type": "link",
          },
        ],
        "pathSegment": "",
        "title": "Group (should be second)",
        "type": "menu",
      },
      {
        "pathSegment": "after",
        "title": "After (should be third)",
        "type": "link",
      },
    ],
    "pathSegment": "parent",
    "title": "Parent title",
    "type": "heading",
  },
]
`;

exports[`generateDocsToc sorts top-level and children items 1`] = `
[
  {
    "children": [
      {
        "pathSegment": "parameters",
        "title": "Parameters (should be first)",
        "type": "link",
      },
      {
        "pathSegment": "decorators",
        "title": "Decorators (should be second)",
        "type": "link",
      },
    ],
    "pathSegment": "writing-stories",
    "title": "How to write stories (should be first)",
    "type": "heading",
  },
  {
    "children": [
      {
        "pathSegment": "autodocs",
        "title": "Automatic documentation and Storybook (should be first)",
        "type": "link",
      },
      {
        "pathSegment": "mdx",
        "title": "MDX (should be second)",
        "type": "link",
      },
    ],
    "pathSegment": "writing-docs",
    "title": "How to document components (should be second)",
    "type": "heading",
  },
  {
    "file": "12-faq.md",
    "pathSegment": "faq",
    "title": "Frequently Asked Questions (should be third)",
    "type": "link",
  },
]
`;
